jpype/_core.py

```diff
@@ -327,17 +331,25 @@ def shutdownJVM():
     restart the JVM after being terminated.
     """
     import threading
+    import jpype.config
     if threading.current_thread() is not threading.main_thread():
         raise RuntimeError("Shutdown must be called from main thread")
-    _jpype.shutdown()
+    if _jpype.isStarted():
+        _jpype.JPypeContext.freeResources = jpype.config.free_resources
+    _jpype.shutdown(jpype.config.destroy_jvm, jpype.config.free_jvm)
 
 
 # In order to shutdown cleanly we need the reference queue stopped
 # otherwise, we can experience a crash if a Java thread is waiting
 # for the GIL.
 def _JTerminate():
     try:
-        _jpype.shutdown()
+        import jpype.config
+        # We are exiting anyway so no need to free resources
+        if _jpype.isStarted():
+            _jpype.JPypeContext.freeResources = False
+        if jpype.config.onexit:
+            _jpype.shutdown(jpype.config.destroy_jvm, False)
```

Comment:
I haven't understood why we wouldn't want to use ``jpype.config.free_jvm`` instead of the ``False`` literal here. Is this a mistake, or intentional?

Reply:
The is no point in freeing if we are 90% of the way to exit.  Is there?